# Chapter 1 Answers: Exercises in 1.9.

## Author
Yu-Ping Lin

## Objectives
In this notebook, I write the answers to the exercises of Chapter 1 in the textbook *Think Python: How to Think Like a Computer Scientist, 3rd Edition* by Allen B. Downey. The questions are adapted from Section 1.9. of the textbook. Please refer to the [official website](https://allendowney.github.io/ThinkPython/chap01.html) or the published versions for the chapter content and the original questions.

## 1.9.1. Ask a virtual assistant

Here are some topics you could ask a virtual assistant about:

### Question 1

Earlier I mentioned bitwise operators but I didn’t explain why the value of ``7 ^ 2`` is ``5``. Try asking "What are the bitwise operators in Python?" or "What is the value of ``7 XOR 2``?"

### Answer

Bitwise operators are operators that compare the binary representations of objects bit-by-bit. The resulting binary representation then defines the result.

For example, ``7 ^ 2`` means ``7 XOR 2``, the **exclusive OR**. It compares the binary representations of ``7``, ``111``, and ``2``, ``10``, then generates a binary that is ``1`` if the bits are different and ``0`` if the bits are the same. The result is ``101``, which turns into the integer ``5``.

In [2]:
7 ^ 2

5

Another interesting example is ``&``, the **bitwise AND**. Anthough it is frequently used in the same way as ``and`` in logical operations, its true effect is to do bitwise comparison, with returns ``1`` and ``0`` for the same and different bits, respectively. For instance, the expression ``5 & 6`` returns ``4`` since it compares ``101`` and ``110`` and returns ``100``.

In [3]:
5 & 6

4

### Question 2

I also mentioned the order of operations. For more details, ask "What is the order of operations in Python?"

### Answer

Below is a table of the orders of operators in Python.

| Precedence | Operators | Description |
|-|-|-|
| 1 | ``()`` | Parentheses (grouping) |
| 2 | ``**`` | Exponentiation |
|3 | ``+x``, ``-x``, ``~x`` | Unary plus, minus, bitwise NOT |
| 4 | ``*``, ``/``, ``//``, ``%`` | Multiplication, division, floor division, modulo |
| 5 | ``+``, ``-`` | Addition and subtraction |
| 6 | ``<<``, ``>>`` | Bitwise shift operators |
| 7 | ``&`` | Bitwise AND |
| 8 | ``^`` | Bitwise XOR |
| 9 | `` ` `` | ` |
| 10 | ``in``, ``not in``, ``is``, ``is not``, ``<``, ``<=``, ``>``, ``>=``, ``!=``, ``==`` | Comparisons |
| 11 | ``not`` | Logical NOT |
| 12 | ``and`` | Logical AND |
| 13 | ``or`` | Logical OR |
| 14 | ``if ... else`` | Conditional expressions (ternary) |
| 15 | ``lambda`` | Lambda expressions |

Interestingly, we can see that the orders of bitwise AND `&` and logical AND `and` are actually different.

### Question 3

The ``round`` function, which we used to round a floating-point number to the nearest integer, can take a second argument. Try asking "What are the arguments of the round function?" or "How do I round pi off to three decimal places?"

### Answer

The syntax of round function is as follows:

``round(number, ndigits = None)``

The arguments are:
1. number: The number to be rounded, whose type can be either ``int`` or ``float``.
2. ndigits: The number of decimals for the round. With the default value ``None``, the function rounds to the nearest integer. If a positive or negative integer ``N`` is given, it rounds to the ``N``-th decimal to the right or left of the decimal point.

Note that if ``number = X.5`` and ``ndigits = None``, the ``round`` function will round it to the nearest even number!

Below are some examples:

In [6]:
round(4.1)

4.1

In [9]:
round(4.9)

5

In [10]:
round(4.5)

4

In [12]:
import math
print(math.pi)
round(math.pi, 3)

3.141592653589793


3.142

### Question 4

There’s one more arithmetic operator I didn’t mention; try asking "What is the modulus operator in Python?"

### Answer

The modulus operator ``%`` returns the remainder of a division. For example, the expression ``a % b`` will return the remainder of ``a`` under the division with ``b``.

Below are some examples:

In [13]:
7 % 3

1

In [14]:
9 % 5

4

## 1.9.2. Exercise

### Question

You might wonder what ``round`` does if a number ends in ``0.5``. The answer is that it sometimes rounds up and sometimes rounds down. Try these examples and see if you can figure out what rule it follows.

In [19]:
round(42.5)

42

In [16]:
round(43.5)

44

### Answer

As mentioned in **Question 3** of **1.9.1. Exercise**, the round function rounds to the nearest even value. Note that this rule applies for any ``ndigits``.

## 1.9.3. Exercise

When you learn about a new feature, you should try it out and make mistakes on purpose. That way, you learn the error messages, and when you see them again, you will know what they mean. It is better to make mistakes now and deliberately than later and accidentally.

### Question 1

You can use a minus sign to make a negative number like ``-2``. What happens if you put a plus sign before a number? What about ``2++2``?

### Answer

In [26]:
2++2

4

``2++2`` means ``2 + +2``, which returns ``4``.

### Question 2

What happens if you have two values with no operator between them, like ``4 2``?

### Answer

In [27]:
4 2

SyntaxError: invalid syntax (436257307.py, line 1)

The execution of this expression returns a ``SyntaxError``, meaning that the structure of this expression is invalid. There must be some operator between two values in an expression.

### Question 3

If you call a function like ``round(42.5)``, what happens if you leave out one or both parentheses?

### Answer

In [28]:
round(42.5

_IncompleteInputError: incomplete input (3923528527.py, line 1)

The expression returns an ``IncompleteInputError, meaning that the input to the function ``round`` is incomplete.

## 1.9.4. Exercise

### Question

Recall that every expression has a value, every value has a type, and we can use the ``type`` function to find the type of any value.

What is the type of the value of the following expressions? Make your best guess for each one, and then use ``type`` to find out.

- ``765``
- ``2.718``
- ``'2 pi'``
- ``abs(-7)``
- ``abs(-7.0)``
- ``abs``
- ``int``
- ``type``

### Answer

See below for the output of the ``type`` function on these values.

In [29]:
print(f"The type of 765 is {type(765)}")
print(f"The type of 2.718 is {type(2.718)}")
print(f"The type of '2 pi' is {type('2 pi')}")
print(f"The type of abs(-7) is {type(abs(-7))}")
print(f"The type of abs(-7.0) is {type(abs(-7.0))}")
print(f"The type of (abs is {type(abs)}")
print(f"The type of int is {type(int)}")
print(f"The type of type is {type(type)}")

The type of 765 is <class 'int'>
The type of 2.718 is <class 'float'>
The type of '2 pi' is <class 'str'>
The type of abs(-7) is <class 'int'>
The type of abs(-7.0) is <class 'float'>
The type of (abs is <class 'builtin_function_or_method'>
The type of int is <class 'type'>
The type of type is <class 'type'>


The last one is especially interesting: The type of ``type`` is ``type``! It can be understood as follows:
1. Everything in Python is an object, including classes.
2. The classes themselves are the instances in the ``type`` metaclass.
3. Since ``type`` is itself a class, its type is also ``type``.

## 1.9.5. Exercise

The following questions give you a chance to practice writing arithmetic expressions. If you already know about variables, you can use them for this exercise. If you don’t, you can do the exercise without them – and then we’ll see them in the next chapter.

(To match the content of this chapter, here we do not use the variables on purpose.)

### Question 1

How many seconds are there in 42 minutes 42 seconds?

### Answer

Since 1 minute is 60 seconds, we can multiply the number of minutes by 60 and add the number of seconds.

In [30]:
42 * 60 + 42

2562

### Question 2

How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.

### Answer

Since 1 mile is 1.61 kilometers, we can divide the number of kilometers by 1.61 to get the number of miles.

In [31]:
10 / 1.61

6.211180124223602

### Question 3

If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace in seconds per mile?

### Answer

In the previous two questions, we have obtained the number of seconds in 42 minutes 42 seconds and the number of miles in 10 kilometers. Our task here is simply divide them.

In [32]:
6.211180124223602 / 2562

0.0024243482139826703

### Question 4

What is your average pace in minutes and seconds per mile?

### Answer

To derive the answer, we first compute the seconds per mile.

In [34]:
2562 / 6.211180124223602

412.482

Then we can use the integer division ``//`` and remainder ``%`` by 60, the number of seconds per minute, to obtain the result.

In [41]:
round(412.482 // 60)

6

In [40]:
round(412.482 % 60, 3)

52.482

So the result is 6 minute 52.482 seconds.

### Question 5

What is your average speed in miles per hour?

### Answer

We can first compute the number of hours in 2562 seconds, with 1 hour equaling 60 minutes in mind.

In [42]:
(2562 / 60) / 60

0.7116666666666667

Then we can divide the number of miles by the number of hours, which gives the average speed in miles per hour.

In [43]:
6.211180124223602 / 0.7116666666666667

8.727653570337614

## This completes the exercises in this Chapter!